readthedocs/core/middleware.py

```diff
@@ -188,34 +193,131 @@ def process_request(self, request):
         return None
 
 
-class FooterNoSessionMiddleware(SessionMiddleware):
+class ReadTheDocsSessionMiddleware(SessionMiddleware):
 
     """
-    Middleware that doesn't create a session on logged out doc views.
+    An overridden session middleware with a few changes.
 
-    This will reduce the size of our session table drastically.
+    - Doesn't create a session on logged out doc views.
+    - Uses a fallback cookie for browsers that reject SameSite=None cookies
+    - Modifies Django's behavior of treating SESSION_COOKIE_SAMESITE=None
+      to mean SameSite unset and instead makes it mean SameSite=None
+
+    This overrides and replaces Django's built-in SessionMiddleware completely.
+    Much of this middleware is duplicated from Django 2.2's SessionMiddleware.
+
+    In Django 3.1, Django will fully support SameSite=None cookies.
+    However, we may still need this middleware to support browsers that reject
+    SameSite=None cookies.
+    https://www.chromium.org/updates/same-site/incompatible-clients
     """
 
+    # Don't set a session cookie on these URLs unless the cookie is already set
     IGNORE_URLS = [
         '/api/v2/footer_html', '/sustainability/view', '/sustainability/click',
     ]
 
+    # This is a fallback cookie for the regular session cookie
+    # It is only used by clients that reject cookies with `SameSite=None`
+    cookie_name_fallback = f"{settings.SESSION_COOKIE_NAME}-samesiteunset"
+
     def process_request(self, request):
         for url in self.IGNORE_URLS:
             if (
                 request.path_info.startswith(url) and
-                settings.SESSION_COOKIE_NAME not in request.COOKIES
+                settings.SESSION_COOKIE_NAME not in request.COOKIES and
+                self.cookie_name_fallback not in request.COOKIES
             ):
                 # Hack request.session otherwise the Authentication middleware complains.
                 request.session = SessionBase()  # create an empty session
                 return
-        super().process_request(request)
+
+        if settings.SESSION_COOKIE_SAMESITE:
+            super().process_request(request)
+        else:
+            if settings.SESSION_COOKIE_NAME in request.COOKIES:
+                session_key = request.COOKIES.get(settings.SESSION_COOKIE_NAME)
+            else:
+                session_key = request.COOKIES.get(self.cookie_name_fallback)
+
+            request.session = self.SessionStore(session_key)
 
     def process_response(self, request, response):
         for url in self.IGNORE_URLS:
             if (
                 request.path_info.startswith(url) and
-                settings.SESSION_COOKIE_NAME not in request.COOKIES
+                settings.SESSION_COOKIE_NAME not in request.COOKIES and
+                self.cookie_name_fallback not in request.COOKIES
             ):
                 return response
-        return super().process_response(request, response)
+
+        try:
```

Comment:
Did we heavily modify the code below, or just small edits to support both cookie names? Would be useful to better understand what exactly we've changed.

Reply:
Mostly it's just small changes. I can add a comment trying to explain it although the places with changes are commented.

Reply:
I added comments about what changed. It isn't just Auth0 making this recommendation. This is recommended directly from the [Google web team](https://web.dev/samesite-cookie-recipes/).